Add Nanosecond Precision Support for Flink-Iceberg Integration

talatuyarer opened the pull request:
This PR fixes the issue where nanosecond precision timestamps were being truncated to millisecond precision when using Apache Flink with Apache Iceberg V3 tables. Now you can actually use those fancy TIMESTAMP(9) and TIMESTAMP_LTZ(9) types without losing precision! 

**The Problem**
When inserting data with nanosecond precision using Flink SQL like:
```sql
INSERT INTO my_table VALUES (TIMESTAMP '2025-01-15 10:30:45.123456789');
```
The data would mysteriously lose precision and come back as:
```
2025-01-15 10:30:45.123456
```

**Core Issues:**
`RowDataWrapper` was converting all timestamps to microseconds regardless of precision
`FlinkParquetWriters` was losing nanosecond precision when writing to Parquet files
`StructRowData` and `RowDataUtil` were truncating precision when reading data back
`FlinkTypeToType` was always mapping `TIMESTAMP(9)` to microsecond Iceberg types (causing Flink to cast to `TIMESTAMP(6)`)

**Data Format Support:**
✅ Parquet: Fixed writers and readers to preserve nanosecond precision
✅ Avro: Already working perfectly (verified with tests)
❌ ORC: Still needs fixing (truncates to microseconds)

If you want to test by yourself.
```sql
-- Create table with nanosecond precision
CREATE TABLE test_table (
    id BIGINT,
    ts TIMESTAMP(9),
    ts_tz TIMESTAMP_LTZ(9)
) WITH (
    'connector' = 'iceberg',
    'catalog-name' = 'hadoop_catalog',
    'warehouse' = 'gs://my-bucket/warehouse',
    'format-version' = 3
);

-- Insert with nanosecond precision
INSERT INTO test_table VALUES 
(1, TIMESTAMP '2025-01-15 10:30:45.123456789', TIMESTAMP '2025-01-15 10:30:45.123456789');

-- Query and verify precision is preserved
SELECT * FROM test_table;
-- Should show: 2025-01-15 10:30:45.123456789 (not truncated!)
```

pvary commented on flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/data/FlinkParquetWriters.java:
```diff
@@ -446,8 +446,9 @@ private TimestampDataWriter(ColumnDescriptor desc) {
     public void write(int repetitionLevel, TimestampData value) {
-      column.writeLong(
-          repetitionLevel, value.getMillisecond() * 1000 + value.getNanoOfMillisecond() / 1000);
+      // For microsecond precision timestamps, convert to microseconds
+      long micros = value.getMillisecond() * 1_000L + value.getNanoOfMillisecond() / 1_000L;
+      column.writeLong(repetitionLevel, micros);
```
Why is this change?

pvary commented on flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/data/FlinkParquetWriters.java:
```diff
@@ -459,8 +460,9 @@ private TimestampNanoDataWriter(ColumnDescriptor desc) {
     @Override
     public void write(int repetitionLevel, TimestampData value) {
-      column.writeLong(
-          repetitionLevel, value.getMillisecond() * 1_000_000L + value.getNanoOfMillisecond());
+      // For nanosecond precision timestamps, convert to nanoseconds
+      long nanos = value.getMillisecond() * 1_000_000L + value.getNanoOfMillisecond();
+      column.writeLong(repetitionLevel, nanos);
```
Could we use the same pattern than the original write for the millisecond timestamp?

pvary commented on flink/v1.20/flink/src/test/java/org/apache/iceberg/flink/data/TestFlinkParquetWriter.java:
```diff
@@ -118,4 +121,98 @@ protected void writeAndValidate(Schema schema, List<Record> expectedData) throws
+
+  /** Test that nanosecond precision timestamps are preserved when writing to Parquet files. */
+  @Test
+  public void testNanosecondTimestampPrecision() throws IOException {
```
Is the nano timestamp supported by any other file format, like Avro, or ORC?

pvary commented on flink/v2.0/flink/src/test/java/org/apache/iceberg/flink/source/reader/TestColumnStatsWatermarkExtractorEndToEnd.java:
```diff
@@ -0,0 +1,202 @@
+          required(1, "id", Types.LongType.get()),
+          required(2, "event_time_ns", Types.TimestampNanoType.withoutZone()),
+          required(3, "event_time_ns_tz", Types.TimestampNanoType.withZone()),
+          required(4, "event_time_us", Types.TimestampType.withoutZone()),
```
question: what does `us` stand for?

talatuyarer replied:
this means µs microsecond symbol. I changed with actual name :)

pvary commented on flink/v2.0/flink/src/test/java/org/apache/iceberg/flink/source/reader/TestColumnStatsWatermarkExtractorEndToEnd.java:
```diff
@@ -0,0 +1,202 @@
+  }
+
+  @TestTemplate
+  public void testWatermarkExtractionWithRealData() throws Exception {
```
What does this test do?
Do we have test which test the actual extraction?

pvary commented on flink/v2.0/flink/src/test/java/org/apache/iceberg/flink/source/reader/TestColumnStatsWatermarkExtractorEndToEnd.java:
```diff
@@ -0,0 +1,202 @@
+  }
+
+  @TestTemplate
+  public void testDataScanningWithNanosecondTimestamps() throws Exception {
```
What is the goal of this test?

talatuyarer replied:
I added this as Javadoc 
   
```
Tests the full write-read cycle for tables with nanosecond precision timestamps. This verifies that data with nanosecond timestamps can be written to files (Parquet/Avro), the data can be read back through Flink's source implementation, and nanosecond precision is preserved during the round-trip. This is an integration test ensuring that the complete data path works correctly with nanosecond timestamps.
```

talatuyarer commented on flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/FlinkTypeToType.java:
```diff
@@ -137,12 +137,20 @@ public Type visit(TimeType timeType) {
+      return Types.TimestampNanoType.withoutZone();
+    } else {
+      return Types.TimestampType.withoutZone();
+    }
```
No This will be behaviour change. Today Flink Iceberg support Timestamp 0 -> 6 and all maps to microsecond.

talatuyarer commented:
Thank you @pvary for reviewing this. As you suggested I removed all other version except 2.1 and also addressed your comments 

- Throw unsupported for Timestamp Precision if it is bigger than 9
- Added Timestamp Nano to DataGenerators 
- Improved Test Documentation 

Please let me know if you see anything missing.

pvary commented on flink/v2.1/flink/src/main/java/org/apache/iceberg/flink/sink/AvroGenericRecordToRowDataMapper.java:
```diff
@@ -41,13 +45,114 @@
+  }
+
+  private RowData convertWithNanos(GenericRecord genericRecord) {
+    RowData baseRowData = (RowData) converter.convert(genericRecord);
+    GenericRowData result = new GenericRowData(baseRowData.getArity());
```
I don't really like this solution. We creating 2 `RowData` objects for a single conversion.
Can we do this with a single conversion?

pvary commented on flink/v2.1/flink/src/main/java/org/apache/iceberg/flink/source/RowDataToAvroGenericRecordConverter.java:
```diff
@@ -41,17 +44,114 @@
+        // For non-timestamp-nanos fields, copy value directly from rowData
+        result.put(i, getFieldValue(rowData, i, field));
+      }
+    }
+    return result;
```
Add a newline after closing every block?
See: https://iceberg.apache.org/contribute/#block-spacing

pvary commented on flink/v2.1/flink/src/main/java/org/apache/iceberg/flink/data/FlinkValueReaders.java:
```diff
@@ -191,10 +191,10 @@ private static class TimestampNanosReader implements ValueReader<TimestampData>
+      long storedValue = decoder.readLong();
+      long mills = Math.floorDiv(storedValue, 1_000_000L);
+      int leftoverNanos = (int) Math.floorMod(storedValue, 1_000_000L);
+      return TimestampData.fromEpochMillis(mills, leftoverNanos);
```
Any relevant change here?
For me this seems like we renamed some of the variables, but nothing has changed

pvary commented on flink/v2.1/flink/src/main/java/org/apache/iceberg/flink/data/FlinkParquetReaders.java:
```diff
@@ -404,8 +404,10 @@ private static class NanosToTimestampReader
+      long nanos = Math.floorMod(value, 1_000_000_000L);
       return TimestampData.fromEpochMillis(
-          Math.floorDiv(value, 1_000_000L), Math.floorMod(value, 1_000_000));
+          mills + Math.floorDiv(nanos, 1_000_000L), (int) Math.floorMod(nanos, 1_000_000L));
```
I don't understand this change.
Why are we precalculating seconds? And then again converting to millis?

pvary commented on flink/v2.1/flink/src/main/java/org/apache/iceberg/flink/data/FlinkValueWriters.java:
```diff
@@ -139,9 +139,9 @@ private static class TimestampNanosWriter implements ValueWriter<TimestampData>
-      encoder.writeLong(nanos);
+      long value =
+          timestampData.getMillisecond() * 1_000_000L + timestampData.getNanoOfMillisecond();
+      encoder.writeLong(value);
```
Unnecessary change again

talatuyarer commented:
@pvary and @Guosmilesmile Next week I will update this PR. I am working on this. Sorrty this took longer than I expect.

pvary commented on flink/v2.1/flink/src/main/java/org/apache/iceberg/flink/data/StructRowData.java:
```diff
@@ -186,8 +186,31 @@ private BigDecimal getDecimalInternal(int pos) {
+      }
+      long mills = Math.floorDiv(nanos, 1_000_000);
+      int leftover = Math.floorMod(nanos, 1_000_000);
+      return TimestampData.fromEpochMillis(mills, leftover);
```
Question: Why not something similar to the line below with `/`, `%` and `*` operators?

pvary commented on flink/v2.1/flink/src/main/java/org/apache/iceberg/flink/sink/AvroGenericRecordToRowDataMapper.java:
```diff
@@ -18,44 +18,119 @@
+          long nanos = (Long) avroValue;
+          long millis = Math.floorDiv(nanos, 1_000_000);
+          int nanosOfMillis = Math.floorMod(nanos, 1_000_000);
+          TimestampData timestampData = TimestampData.fromEpochMillis(millis, nanosOfMillis);
```
Same question as above. Any specific reason to use math?

pvary commented on flink/v2.1/flink/src/main/java/org/apache/iceberg/flink/sink/AvroGenericRecordToRowDataMapper.java:
```diff
@@ -18,44 +18,119 @@
+    }
+    // Use FieldGetter to extract the field value
+    LogicalType fieldType = rowType.getTypeAt(pos);
+    RowData.FieldGetter fieldGetter = RowData.createFieldGetter(fieldType, pos);
```
Would it be possible to create the fieldgetters in the constructor when there is a nano type present?
It would save quite a bit of time if we don't recreate them every time.

pvary commented on flink/v2.1/flink/src/main/java/org/apache/iceberg/flink/source/RowDataToAvroGenericRecordConverter.java:
```diff
@@ -41,30 +48,111 @@
+        } else {
+          LogicalType fieldType = converterRowType.getTypeAt(i);
+          RowData.FieldGetter fieldGetter = RowData.createFieldGetter(fieldType, i);
+          processedRowData.setField(i, fieldGetter.getFieldOrNull(rowData));
```
Don't create the fieldGetters every time

pvary commented on flink/v2.1/flink/src/main/java/org/apache/iceberg/flink/source/RowDataToAvroGenericRecordConverter.java:
```diff
@@ -41,30 +48,116 @@
-    return new RowDataToAvroGenericRecordConverter(rowType, avroSchema);
+
+    // Fix up timestamp-nanos fields: Flink's RowDataToAvroConverters expects Long (nanoseconds)
+    // for timestamp-nanos logical type, but our RowData has TimESTAMP(9) with TimestampData.
```
nit: `TimESTAMP`

mxm commented on flink/v2.1/flink/src/main/java/org/apache/iceberg/flink/FlinkTypeToType.java:
```diff
@@ -137,12 +137,30 @@ public Type visit(TimeType timeType) {
+    }
+    if (precision == 9) {
+      return Types.TimestampNanoType.withoutZone();
+    } else {
```
Yes, they are possible. Not if this Flink type has been constructed from an Iceberg table, but there are other ways in Flink to create a timestamp with precision < 9.

mxm commented on flink/v2.1/flink/src/main/java/org/apache/iceberg/flink/source/RowDataToAvroGenericRecordConverter.java:
```diff
@@ -41,30 +48,116 @@
+
+    // Detect timestamp-nanos fields and create converter RowType with BIGINT for those fields
+    List<LogicalType> converterFields = Lists.newArrayList();
+    Set<Integer> timestampNanosFieldIndices = new java.util.HashSet<>();
```
NIT: 
```suggestion
    Set<Integer> timestampNanosFieldIndices = Sets.newHashSet();
```

talatuyarer commented:
Yes Sorry for delaying. Let me update PR this week @pvary
